src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,445 @@
+    (ensures tr1 == tr2)
+// TODO:
+// It is not clear if this SMTPat is ever practically useful, but it seems like it could be in principle
+    [SMTPat (tr1 == tr2); SMTPat (tr1 <$ tr2)]
+  = prefix_prefix_eq tr1 tr2 (length tr1)
+
+let grows_cases (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures tr1 == tr2 \/ (
+      Snoc? tr2 /\ (
+      let Snoc hd _ = tr2 in
+      tr1 <$ hd
+    )))
+  = if length tr1 = length tr2
+    then grows_full_eq tr1 tr2
+    else begin
+      let Snoc hd e = tr2 in
+      calc (<$) {
+        tr1;
+        == {}
+        prefix tr1 (length tr1);
+        <$ { prefix_prefix_grows tr1 tr2 (length tr1) (length hd) }
+        prefix tr2 (length hd);
+        == { prefix_prefix_eq hd tr2 (length hd) }
+        prefix hd (length hd);
+        == { }
+        hd;
+      }
+    end
+
+
+(*** General utility functions for working with traces ***)
+
+// This function is not terribly important, but we often want to check if we are at
+// the end of a trace, and hiding the off-by-one between length and the last index
+// inside a function makes things look cleaner.
+let last_ts (#label_t:Type) (tr:trace_ label_t{Snoc? tr}) : timestamp = length tr - 1
+
+// Now that trace entries are abstracted over label types, we cannot compare
+// trace entries for equality with =. This function implements trace equality
+// when the label type is an eqtype (as for unit, for instance).
+let trace_entry_equal (#label_t:eqtype) (e1 e2:trace_entry_ label_t)
+  : bool
+  = match (e1, e2) with
+    | MsgSent b1, MsgSent b2 -> b1 = b2
+    | RandGen u1 l1 i1, RandGen u2 l2 i2 -> u1 = u2 && l1 = l2 && i1 = i2
+    | Corrupt ts1, Corrupt ts2 -> ts1 = ts2
+    | SetState p1 sid1 b1, SetState p2 sid2 b2 -> p1 = p2 && sid1 = sid2 && b1 = b2
+    | Event p1 t1 b1, Event p2 t2 b2 -> p1 = p2 && t1 = t2 && b1 = b2
+    | _ -> false
+
+// By forgetting labels, we can work with a label-ignoring equivalence on entries
+let trace_entry_equiv (#label_t:Type) (e1 e2:trace_entry_ label_t)
+  : bool
+  = trace_entry_equal #unit (fmap_trace_entry forget_label e1) (fmap_trace_entry forget_label e2)
+
+let trace_entry_equal_is_eq (#label_t:eqtype) (e1 e2:trace_entry_ label_t)
+  : Lemma
+    (ensures trace_entry_equal e1 e2 <==> e1 == e2)
+    [SMTPat (trace_entry_equal e1 e2)]
+  = ()
+
+// trace_entry_equiv is an equivalence relation
+let trace_entry_equiv_reflexive (#label_t:Type) (e:trace_entry_ label_t)
+  : Lemma
+    (ensures trace_entry_equiv e e)
+    [SMTPat (trace_entry_equiv e e)]
+  = ()
+
+let trace_entry_equiv_symmetric (#label_t:Type) (e1 e2:trace_entry_ label_t)
+  : Lemma
+    (requires trace_entry_equiv e1 e2)
+    (ensures trace_entry_equiv e2 e1)
+  = ()
+
+let trace_entry_equiv_transitive (#label_t:Type) (e1 e2 e3:trace_entry_ label_t)
+  : Lemma
+    (requires
+       trace_entry_equiv e1 e2 /\
+       trace_entry_equiv e2 e3
+    )
+    (ensures trace_entry_equiv e1 e3)
+  = ()
```

Comment:
Are these lemmas we are expected to use at some point, or are there sanity checks?
Usually proofs about transitivity / symmetry etc are really a pain, and here it looks like we could let the SMT solver do the job here.

Reply:
I don't expect most of these to be needed, though I could see reflexivity potentially coming up at some point. As you note, though, the SMT solver seems to handle this without a problem, and should continue to do so as long as trace_entry_equiv is not opaque.